### PR TITLE
- Let DotProduct,Wand and WeightedSet be Term nodes in the query tree…

### PR DESCRIPTION
**Pull request by baldersheim**
… as they really are.

  That restricts the nodes to what they can really do and makes them significantly cheaper.
- In addition type conversion of numeric terms is delayed to when it is necessary.
   And as next step they can be avoided completely.

@havardpe PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/query/tree/queryreplicator.h [hunk 1]
```diff
@@ -78,26 +78,32 @@ class QueryReplicator : private QueryVisitor {
         visitNodes(node.getChildren());
     }
 
+    void replicateMultiTerm(const MultiTerm &original, MultiTerm & replica) {
+        for (uint32_t i(0); i < original.getNumTerms(); i++) {
+            auto v = original.getAsString(i);
+            replica.addTerm(v.first, v.second);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/query/tree/queryreplicator.h:84]: this makes the replica different from the original in some cases

File: searchlib/src/vespa/searchlib/query/tree/stackdumpquerycreator.h [hunk 2]
```diff
@@ -45,8 +47,26 @@ class StackDumpQueryCreator {
         return builder.build();
     }
 
-private: 
-    static Term * createQueryTerm(search::SimpleQueryStackDumpIterator &queryStack, QueryBuilder<NodeTypes> & builder, vespalib::stringref & pureTermView) {
+private:
+    static void populateMultiTerm(search::SimpleQueryStackDumpIterator &queryStack, QueryBuilderBase & builder, MultiTerm & mt) {
+        for (uint32_t i(0); i < mt.getNumTerms(); i++) {
+            queryStack.next();
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/query/tree/stackdumpquerycreator.h:53]: need to check for errors and end-of-stack (also, we might want to roll back to a deterministic error state in the next() function in case of en exception, just in case someone does something like this in the future...)

File: searchcore/src/tests/proton/matching/query_test.cpp [hunk 3]
```diff
@@ -164,6 +164,17 @@ fef_test::IndexEnvironment plain_index_env;
 fef_test::IndexEnvironment resolved_index_env;
 fef_test::IndexEnvironment attribute_index_env;
 
+vespalib::string
+termAsString(const search::query::Range &term) {
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/query_test.cpp:168]: Why are these here? (to avoid updating the test to use the new API?)

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp [hunk 4]
```diff
@@ -1,3 +1,32 @@
...
+namespace {
+class StringAsKey final : public IDocumentWeightAttribute::LookupKey {
+public:
+    StringAsKey(vespalib::stringref key)
+            : _key(key)
+    { }
+
+    vespalib::stringref asString() const override { return _key; }
+private:
+    vespalib::string _key;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp:16]: you probably do not need a copy here

File: searchlib/src/vespa/searchlib/query/tree/termnodes.cpp [hunk 5]
```diff
@@ -1,26 +1,111 @@
...
+}
+
+class StringTermVector final : public MultiTerm::TermVector {
+public:
+    StringTermVector(uint32_t sz) : _terms() { _terms.reserve(sz); }
+    void addTerm(stringref term, Weight weight) override {
+        _terms.emplace_back(term, weight);
+    }
+    void addTerm(int64_t, Weight) override {
+        badType("string", "int64_t");
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/query/tree/termnodes.cpp:42]: this is a very well hidden restriction of the query stack format... how is this modeled in Java? is the term data type annotated on the parent multi-term node?

File: searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp [hunk 6]
```diff
@@ -66,12 +66,12 @@ template <typename WS, typename NODE>
...
 CreateBlueprintVisitorHelper::createWeightedSet(std::unique_ptr<WS> bp, NODE &n) {
     FieldSpecList fields;
-    for (size_t i = 0; i < n.getChildren().size(); ++i) {
+    for (size_t i = 0; i < n.getNumTerms(); ++i) {
         fields.clear();
         fields.add(bp->getNextChildField(_field));
-        const query::Node &node = *n.getChildren()[i];
-        uint32_t weight = getWeightFromNode(node).percent();
-        bp->addTerm(_searchable.createBlueprint(_requestContext, fields, node), weight);
+        const auto & term = n.getAsString(i);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp:72]: when you have a const reference to a temporary object that contains a string reference to a mutable scratch buffer things become a bit more complicated than they seem...

Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/create_blueprint_visitor_helper.cpp:72]: Fixed

File: searchcore/src/tests/proton/matching/query_test.cpp (hunk 3 above)
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/query_test.cpp:168]: Only used in test.

File: searchlib/src/vespa/searchlib/query/tree/stackdumpquerycreator.h (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/tree/stackdumpquerycreator.h:53]: Fixed

File: searchlib/src/vespa/searchlib/query/tree/queryreplicator.h (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/tree/queryreplicator.h:84]: Fixed by introducing a type.

File: searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp [hunk 7]
```diff
@@ -108,10 +108,10 @@ MultiValueStringPostingAttributeT<B, T>::DocumentWeightAttributeAdapter::get_dic
 
 template <typename B, typename T>
 IDocumentWeightAttribute::LookupResult
-MultiValueStringPostingAttributeT<B, T>::DocumentWeightAttributeAdapter::lookup(const vespalib::string &term, vespalib::datastore::EntryRef dictionary_snapshot) const
+MultiValueStringPostingAttributeT<B, T>::DocumentWeightAttributeAdapter::lookup(const LookupKey & key, vespalib::datastore::EntryRef dictionary_snapshot) const
 {
     const IEnumStoreDictionary& dictionary = self._enumStore.get_dictionary();
-    auto comp = self._enumStore.make_folded_comparator(term.c_str());
+    auto comp = self._enumStore.make_folded_comparator(key.asString().data());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp:114]: Ensured it is zero terminated at the source.
However the zero termination restriction should be lifted on the comparators in the enumstore.

File: searchlib/src/vespa/searchlib/query/tree/termnodes.cpp (hunk 5 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/tree/termnodes.cpp]: Restriction lifted

**Comment by baldersheim** [comment]
I have added some basic testing of MultiTerm too.

File: searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp (hunk 7 above)
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp:114]: I still think this is dangerous, since it makes the LookupKey API dishonest and using it in a natural way could cause bad results.

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h [hunk 8]
```diff
@@ -3,17 +3,20 @@
...
 
-namespace query { class Node; }
-
 using DocumentWeightIterator = attribute::PostingListTraits<int32_t>::const_iterator;
 
 struct IDocumentWeightAttribute
 {
+    struct LookupKey {
+        virtual ~LookupKey() = default;
+        virtual vespalib::stringref asString() const = 0;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h:16]: this should probably have a warning sign about the return value being 0-terminated

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h [hunk 9]
```diff
@@ -25,7 +28,8 @@ struct IDocumentWeightAttribute
             : posting_idx(posting_idx_in), posting_size(posting_size_in), min_weight(min_weight_in), max_weight(max_weight_in), enum_idx(enum_idx_in) {}
     };
     virtual vespalib::datastore::EntryRef get_dictionary_snapshot() const = 0;
-    virtual LookupResult lookup(const vespalib::string &term, vespalib::datastore::EntryRef dictionary_snapshot) const = 0;
+    virtual LookupResult lookup(const LookupKey & key, vespalib::datastore::EntryRef dictionary_snapshot) const = 0;
+    LookupResult lookup(vespalib::stringref term, vespalib::datastore::EntryRef dictionary_snapshot) const;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h:32]: this should probably have a warning sign about the term needing to be 0-terminated

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp [hunk 10]
```diff
@@ -13,7 +13,7 @@ class StringAsKey final : public IDocumentWeightAttribute::LookupKey {
 
     vespalib::stringref asString() const override { return _key; }
 private:
-    vespalib::string _key;
+    vespalib::stringref _key;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp:16]: it might actually be better to keep the string, to avoid the 0-terminated stringref requirement to escape the LookupKey API and into the attribute API...

File: searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp (hunk 7 above)
Comment (havardpe) [reply searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp:114]: I suggest adding an assert that verifies the 0-termination to make sure it fails in a deterministic way

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h (hunk 8 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h:16]: Note added.

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h (hunk 9 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.h:32]: Adde a note being only used in tests. Zero termination handled on the inside.

File: searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp (hunk 7 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/multistringpostattribute.hpp:114]: Assert added

File: searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp (hunk 10 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/i_document_weight_attribute.cpp:16]: done

**Comment by baldersheim** [comment]
Comments addressed in https://github.com/vespa-engine/vespa/pull/17251
